lib/src/optimism/batcher.rs

```diff
@@ -0,0 +1,458 @@
+    pub timestamp: u64,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct Epoch {
+    pub number: BlockNumber,
+    pub hash: B256,
+    pub timestamp: u64,
+    pub base_fee_per_gas: U256,
+    pub deposits: Vec<Transaction<OptimismTxEssence>>,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct State {
+    pub current_l1_block_number: BlockNumber,
+    pub current_l1_block_hash: BlockHash,
+    pub safe_head: BlockInfo,
+    pub epoch: Epoch,
+    pub op_epoch_queue: VecDeque<Epoch>,
+    pub next_epoch: Option<Epoch>,
+}
+
+impl State {
+    pub fn new(
+        current_l1_block_number: BlockNumber,
+        current_l1_block_hash: BlockHash,
+        safe_head: BlockInfo,
+        epoch: Epoch,
+    ) -> Self {
+        State {
+            current_l1_block_number,
+            current_l1_block_hash,
+            safe_head,
+            epoch,
+            op_epoch_queue: VecDeque::new(),
+            next_epoch: None,
+        }
+    }
+
+    pub fn do_next_epoch(&mut self) -> anyhow::Result<()> {
+        self.epoch = self.next_epoch.take().expect("No next epoch!");
+        self.deque_next_epoch_if_none()?;
+        Ok(())
+    }
+
+    pub fn push_epoch(&mut self, epoch: Epoch) -> anyhow::Result<()> {
+        self.op_epoch_queue.push_back(epoch);
+        self.deque_next_epoch_if_none()?;
+        Ok(())
+    }
+
+    fn deque_next_epoch_if_none(&mut self) -> anyhow::Result<()> {
+        if self.next_epoch.is_none() {
+            while let Some(next_epoch) = self.op_epoch_queue.pop_front() {
+                if next_epoch.number <= self.epoch.number {
+                    continue;
+                } else if next_epoch.number == self.epoch.number + 1 {
+                    self.next_epoch = Some(next_epoch);
+                    break;
+                } else {
+                    anyhow::bail!("Epoch gap!");
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum BatchStatus {
+    Drop,
+    Accept,
+    Undecided,
+    Future,
+}
+
+pub struct Batcher {
+    batches: BinaryHeap<Reverse<Batch>>,
+    batcher_channel: BatcherChannels,
+    pub state: State,
+    pub config: ChainConfig,
+}
+
+impl Batcher {
+    pub fn new(
+        config: ChainConfig,
+        op_head: BlockInfo,
+        eth_block: &BlockInput<EthereumTxEssence>,
+    ) -> Result<Batcher> {
+        let eth_block_hash = eth_block.block_header.hash();
+
+        let batcher_channel = BatcherChannels::new(&config);
+
+        let state = State::new(
+            eth_block.block_header.number,
+            eth_block_hash,
+            op_head,
+            Epoch {
+                number: eth_block.block_header.number,
+                hash: eth_block_hash,
+                timestamp: eth_block.block_header.timestamp.try_into().unwrap(),
+                base_fee_per_gas: eth_block.block_header.base_fee_per_gas,
+                deposits: deposits::extract_transactions(&config, eth_block)?,
```

Comment:
I don't think we actually need to extract deposits here.

`Batcher::new()` is always given the L1 block corresponding to the Op head's epoch. In particular, if that block contains any deposits, they would have already been processed either in the Op head or one of its predecessors.